### ISSUE TITLE
bug (data): Document_modified from nested collection incorrectly merged in the data reducer

### ISSUE THREAD

**Issue by compojoom** [prescottprue/redux-firestore#88]
**Do you want to request a *feature* or report a *bug*?**  
bug

**What is the current behavior?**
When a document gets updated and the DOCUMENT_MODIFIED is triggered it isn't merging the data correctly in the data reducer.

**What is the expected behavior?**
We have a listener on companies/{companyid}/users
When a user object gets changed a DOCUMENT_MODIFIED action gets dispatched, but instead of merging the new user inside of companies/{companyid}/users/{userid} -> the document is being inserted in /companies/{userId}

**Which versions of dependencies, and which browser are affected by this issue? Did this work in previous versions or setups?**
I'm using 0.4.2, but I see that the document_modified hasn't changed in 0.4.3

**Comment by compojoom** [comment]
@prescottprue I hope the following info can help you. I looked at this a little further and as usual the problem looks bigger than originally anticipated. 

Originally I thought that you are constructing the wrong path for merging, which is still the case, but it's not the fault of the pathFromMeta function in utils/reducers.js

The problem comes from the docChangeEvent function. There we do this `meta: { ...originalMeta, doc: change.doc.id },` which is fine when you don't have a nested collection, but if you do, then you override the original doc id. 

For my case I made a hack inside of docChange
```
function docChangeEvent(change, originalMeta = {}) {
  const changes = originalMeta.subcollections ? {subcollections: [{...originalMeta.subcollections[0], doc: change.doc.id}]}
    : {doc: change.doc.id}

  return {
    type: changeTypeToEventType[change.type] || actionTypes.DOCUMENT_MODIFIED,
    meta: { ...originalMeta, ...changes },
    payload: {
      data: change.doc.data(),
      ordered: { oldIndex: change.oldIndex, newIndex: change.newIndex },
    },
  };
}
```

This way if I have one subcollection, I'm adding the document at the right position, but that is not a solution that would work for any collection depth. 

I'm not 100% sure how this work.

if I listen on collection1/doc1/subcollection1/subdoc1/subcollection2/
I'm going to be notified with DOCUMENT_MODIFIED for the documents inside subcollection 2, right? I won't be notified if subdoc1 changes, since I'm not directly listening on this path.

**Comment by jwbrew** [comment]
I've been experiencing the same issue today. `DOCUMENT_ADDED` metadata didn't include the id of the subocollection item being created, therefore `pathFromMeta` was constructing the wrong path and inserting it at the parent level.

Weirdly occasionally it would work when the store was populated with a `LISTENER_RESPONSE` instead of `DOCUMENT_ADDED`. 

Wondering whether it has something to do with recent Firestore updates (my version is 0.3.7)?

Anyway, my workaround for the time being is as so: 

1) In `docChangeEvent` extract the path directly from the changed doc and add to meta:  
```
meta: { ...originalMeta, doc: change.doc.id, path: change.doc.ref.path }
```

2) Check for `path` key in `pathFromMeta` and use that if available
```
  if (storeAs) {
    return doc ? `${storeAs}.${doc}` : storeAs;
  }
  if (meta.path) {
    return meta.path.replace(/\//g, ".");
  }
  if (!collection) {
    throw new Error("Collection is required to construct reducer path.");
  }
```

Happy to submit a PR @prescottprue if you think appropriate 😀 

Thanks for the awesome lib!

**Comment by compojoom** [comment]
and here I was trying to traverse nested arrays and update the last subcollection with a docid.

@jwbrew - this looks like a cool solution to me. I tested it and it works in most cases. 

Unfortunately there is still this problem though https://github.com/prescottprue/react-redux-firebase/issues/427

if the doc you are updating has the syntax of example@example.com the updated object will end up example@example: {com: {...}} :(

**Comment by compojoom** [comment]
the issue here is a "problem" with the way lodash treats the dot. The workaround suggested on the lodash issues list is to use array of paths instead of a string separated with dot.
So if we do: 
```
if (meta.path) {
    return meta.path.split(`/`);
  }
```
then updating a document id example@example.com works fine.

Maybe we should change pathfromMeta to return an array? lodash get and set should work just find with it. 

@prescottprue - what do you think?

**Comment by jwbrew** [comment]
Good catch on the syntax @compojoom. Array sounds like a much better solution.

**Comment by prescottprue** [comment]
Commented on there, but would love to get this into `0.5.0`.

Been working on adding support to the orderedReducer for using `storeAs` with subcollections, so haven't gotten into this beyond looking at @compojoom commit.

**Comment by prescottprue** [comment]
@compojoom Really like the usage of path, so I went ahead and added your solution to the `v0.5.0` branch. Going to wait until we understand the impact on `storeAs` before publishing to latest.

**Comment by prescottprue** [comment]
Fix proposed by @jwbrew included as part of [`v0.5.0` release](https://github.com/prescottprue/redux-firestore/releases/tag/v0.5.0).

@jwbrew and @compojoom thanks for coming up with the fix. Reach out if it doesn't work as you expect.
